From 3975045cc8d29f60477a3efa4739ab9a5cada4db Mon Sep 17 00:00:00 2001
From: Lu Yahan <yahan@iscas.ac.cn>
Date: Fri, 10 Jun 2022 17:27:10 +0800
Subject: [PATCH] [riscv64] Fix mem segment fault in simulator

Run mem in simulator may be occurs segments fault. This cl to fix it.
If value of reg is small int, it should be smi.

Change-Id: I60b4eb8c959bc9f86ae28718ff6dd54ecf40a6ec
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/3698757
Auto-Submit: Yahan Lu <yahan@iscas.ac.cn>
Commit-Queue: ji qiu <qiuji@iscas.ac.cn>
Reviewed-by: ji qiu <qiuji@iscas.ac.cn>
Cr-Commit-Position: refs/heads/main@{#81061}
---
 src/execution/riscv64/simulator-riscv64.cc | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/execution/riscv64/simulator-riscv64.cc b/src/execution/riscv64/simulator-riscv64.cc
index 8ae38ca6ae..203381b2d7 100644
--- a/src/execution/riscv64/simulator-riscv64.cc
+++ b/src/execution/riscv64/simulator-riscv64.cc
@@ -7005,8 +7005,9 @@ void Simulator::CallInternal(Address entry) {
   int64_t sp_val = get_register(sp);
 
   // Set up the callee-saved registers with a known value. To be able to check
-  // that they are preserved properly across JS execution.
-  int64_t callee_saved_value = icount_;
+  // that they are preserved properly across JS execution. If this value is
+  // small int, it should be SMI.
+  int64_t callee_saved_value = icount_ << (kSmiTagSize + kSmiShiftSize);
   set_register(s0, callee_saved_value);
   set_register(s1, callee_saved_value);
   set_register(s2, callee_saved_value);
-- 
2.35.1

